libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt

```diff
@@ -600,8 +640,11 @@ object MapboxRouteLineUtils {
                         geoIndex = stepIntersection.geometryIndex()
                     }
                 } else {
-                    if (geoIndex != null && stepIntersection.geometryIndex() != null) {
-                        ranges.add(IntRange(geoIndex!!, stepIntersection.geometryIndex()!!))
+                    ifNonNull(
+                        geoIndex,
+                        stepIntersection.geometryIndex()
+                    ) { startGeoIndex, intersectionGeoIndex ->
+                        ranges.add(IntRange(startGeoIndex, intersectionGeoIndex - 1))
```

Comment:
To get the restricted data I look at the intersections in the leg steps. In each Intersection I look at the classes array to see if the word "restricted" is in it.  If it's found then I look at the `geometryIndex` for that intersection.  Later in the `getRestrictedRouteLegRanges` I return a list of `IntRange` objects.  The `IntRange` is inclusive of the start value to the end value so in the code you refer to I use intersectionGeoIndex - 1 to close the range so that it contains only the range designated by the `geometryIndex` items for intersections that had 'restricted'.  So if there is only one Intersection with 'restricted' the range will be from X to X. If instead there are two (or more) consecutive intersections that are restricted the range will be from X to Y.  These `geometryIndex` values correspond to the `distance` array in the leg annotation.  While looping through the distances array I check if the current array index in the loop is present in the restriction ranges previously calculated.

Reply:
I"ll cut a ticket to refactor this since it's more complex than necessary.

Reply:
>  So if there is only one Intersection with 'restricted' the range will be from X to X

Looking at the API docs:
> intersections.classes [...] An array of strings signifying the classes of **the road exiting the intersection**.

So if X is an intersection that does have a `restricted` class, and Y does not, the range should be from X to Y inclusive. Meaning we should draw a line between the X and Y. And I think I'm starting to understand how it works - because of the distances array. Since the array element at index X describes the distance from X to Y. That's why our range actually needs to be _not inclusive_ because the array of distances is forward-looking in a sense.

And it looks like in the traffic implementation we're actually approaching this from the other direction, we're looking at the _previous_ element in the distance array to work around this forward-looking behavior - https://github.com/mapbox/mapbox-navigation-android/blob/e1824f69cbf4c98d13a831e98cabed4f70003986/libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtils.kt#L553

This would mean that both implementations should be correct and there hopefully are no bugs lingering.

Reply:
Right. In the latest iteration for the restricted road sections the code really shouldn't be concerned with ranges but instead should just return the intersection geometry indexes that have restrictions because that's the data that's actually needed.